File: EpnSampleNativePlugin/EditorPlugin.cpp
#include <Urho3D/Engine/PluginApplication.h>
#include "../GameCore/GameCore.h"


using namespace Urho3D;


class EditorPlugin : public PluginApplication
{
    URHO3D_OBJECT(EditorPlugin, PluginApplication);
public:
    explicit EditorPlugin(Context* context) : PluginApplication(context) { }

    void OnLoad() override
    {
        // Called when editor is loading plugin. Do all plugin setup here.
        RegisterCoreLibrary(context_);
    }

    void OnUnload() override
    {
        // Editor is exiting or plugin is being reloaded. Undo all the changes performed in OnLoad(). After this method
        // exits this library will be unloaded from memory. Failure to clean editor state from any objects provided by
        // this library will result in a crash.
        UnregisterCoreLibrary(context_);
    }
};

// Define entry point of the plugin.
URHO3D_DEFINE_PLUGIN_MAIN(EditorPlugin);

File: EpmSampleManagedPlugin/EditorPlugin.cs
using Urho3DNet;
using ImGuiNet;

// Name of class inheriting PluginApplication and enclosing namespace must match file name of the library.
namespace EpmSampleManagedPlugin
{
    // Every plugin must contain a class which inherits form PluginApplication.
    public class EpmSampleManagedPlugin : PluginApplication
    {
        public EpmSampleManagedPlugin(Context context) : base(context) { }

        // Plugin is being loaded. Called immediately after PluginMain(). Do plugin setup here.
        public override void Load()
        {
            SubscribeToEvent("EditorApplicationMenu", args =>
            {
                if (ImGui.BeginMenu("Managed Plugin Example"))
                {
                    if (ImGui.MenuItem("Greetings!"))
                    {
                        // Log.Write(Log.LogInfo, "Managed editor plugin says hello!");
                    }
                    ImGui.EndMenu();
                }
            });
        }

        // Plugin is being unloaded. Unregister factories, unsubscribe from events, and undo all modifications to editor
        // state. When this call returns this assembly will be unloaded from memory. Failure to comply with this
        // requirement will lead to crashes!
        public override void Unload()
        {
            UnsubscribeFromAllEvents();
        }

        // Main entry point of plugin. Called on plugin load. Most of the time this method should only return a new
        // instance of PluginApplication subclass.
        public static EpmSampleManagedPlugin PluginMain(Context context)
        {
            return new EpmSampleManagedPlugin(context);
        }
    }
}
